core/pkg/server/server.go

```diff
@@ -49,169 +41,157 @@ type Server struct {
 	<-s.serverLifetimeCtx.Done()
 	slog.Info("server is shutting down")
 
 	// Stop accepting new connections.
-	if err := s.listener.Close(); err != nil {
+	if err := listener.Close(); err != nil {
 		slog.Error("failed to Close listener", "error", err)
 	}
 
 	// Wait for asynchronous work to finish.
 	s.wg.Wait()
 
 	slog.Info("server is closed")
+	return nil
 }
 
-func (s *Server) serve() {
-	slog.Info("server is running", "addr", s.listener.Addr())
+// acceptConnections accepts incoming connections on the listener.
+//
+// It blocks until the listener is closed or an error is encountered.
+func (s *Server) acceptConnections(listener net.Listener) {
+	slog.Info("server: accepting connections", "addr", listener.Addr())
 
-	streamMux := stream.NewStreamMux()
-	gpuResourceManager := monitor.NewGPUResourceManager(s.enableDCGMProfiling)
-
-	// Run a separate goroutine to handle incoming connections
 	for {
-		conn, err := s.listener.Accept()
-		if err != nil {
-			select {
-			case <-s.serverLifetimeCtx.Done():
-				slog.Debug("server shutting down...")
-				return
-			default:
-				slog.Error("failed to accept conn.", "error", err)
-			}
-		} else {
-			s.wg.Add(1)
-			go func() {
-				NewConnection(
-					s.serverLifetimeCtx,
-					s.stopServer,
-					ConnectionParams{
-						Conn:               conn,
-						StreamMux:          streamMux,
-						GPUResourceManager: gpuResourceManager,
-						SentryClient:       s.sentryClient,
-						Commit:             s.commit,
-						LoggerPath:         s.loggerPath,
-						LogLevel:           s.logLevel,
-					},
-				).ManageConnectionData()
-
-				s.wg.Done()
-			}()
-		}
-	}
-}
-
-func writePortFile(portFile string, port int) error {
-	tempFile := fmt.Sprintf("%s.tmp", portFile)
-	f, err := os.Create(tempFile)
-	if err != nil {
-		err = fmt.Errorf("fail create temp file: %w", err)
-		return err
-	}
-
-	if _, err = fmt.Fprintf(f, "sock=%d\n", port); err != nil {
-		err = fmt.Errorf("fail write port: %w", err)
-		return err
-	}
+		conn, err := listener.Accept()
 
-	if _, err = f.WriteString("EOF"); err != nil {
-		err = fmt.Errorf("fail write EOF: %w", err)
-		return err
-	}
+		if errors.Is(err, net.ErrClosed) {
+			slog.Info("server: listener closed", "addr", listener.Addr())
+			return
+		}
 
-	if err = f.Sync(); err != nil {
-		err = fmt.Errorf("fail sync: %w", err)
-		return err
-	}
+		// We shut down on any error, even though some may be recoverable.
+		// From the accept(2) manual pages, these include:
+		//
+		//   - EMFILE, ENFILE (out of file descriptors)
+		//   - ENOBUFS, ENOMEM (out of memory)
+		//
+		// Maybe ECONNABORTED as well.
+		//
+		// Though in theory these are recoverable, we don't have a clean way
+		// to distinguish them from other errors in Go, and we want to avoid
+		// busy-looping in case Accept() immediately returns an error
+		// each time it is called.
+		if err != nil {
+			slog.Error("server: failed to accept connection", "error", err)
+			s.stopServer()
+			return
+		}
```

Comment:
is this logic the same, i'm trying to remember why we had the logic structured the other way before

Reply:
Updated this to be more lenient (I thought we couldn't test for specific system errors, but they exist in `syscall`). The original code retried all errors; this PR retries some specific ones picked from the list on the [accept() manpage](https://man7.org/linux/man-pages/man2/accept.2.html#ERRORS) and shuts down on unexpected problems.

One interesting one I don't retry is `EPERM`. I'm not sure if this one happens immediately or when a connection is attempted (so whether it depends on the peer or not). I don't retry it in case it isn't recoverable (such as if it doesn't depend on the peer).